File: web/src/api/planImports.ts
import request from "./request";
import type {
  PlanImportDraftRead,
  PlanImportStructuredPayload,
  PlanImportWorkoutItem,
} from "@/types/models";

export const PLAN_IMPORT_TEMPLATE_FILENAME = "plan-import-template.xlsx";

export function downloadPlanImportTemplate() {
  return request.get<Blob>("/plan-imports/template", {
    responseType: "blob",
  });
}

export function createStructuredPlanImport(payload: PlanImportStructuredPayload, idempotencyKey?: string) {
  return request.post<PlanImportDraftRead>("/plan-imports/structured", payload, {
    headers: idempotencyKey ? { "Idempotency-Key": idempotencyKey } : undefined,
  });
}

export function uploadPlanImportFile(file: File, payload: Omit<PlanImportStructuredPayload, "workouts">) {
  const formData = new FormData();
  formData.append("file", file);
  formData.append("source", payload.source || "file");
  formData.append("anchor_strategy", payload.anchor_strategy);
  formData.append("merge_strategy", payload.merge_strategy);
  formData.append("timezone", payload.timezone || "Asia/Shanghai");
  if (payload.client_request_id) formData.append("client_request_id", payload.client_request_id);
  if (payload.effective_date) formData.append("effective_date", payload.effective_date);
  if (payload.target_cycle_id) formData.append("target_cycle_id", String(payload.target_cycle_id));
  if (payload.target_block_id) formData.append("target_block_id", String(payload.target_block_id));

  return request.post<PlanImportDraftRead>("/plan-imports/file", formData, {
    headers: {
      "Content-Type": "multipart/form-data",
    },
  });
}

export function getPlanImport(importId: number) {
  return request.get<PlanImportDraftRead>(`/plan-imports/${importId}`);
}

export function updatePlanImportItem(importId: number, itemId: number, normalizedItem: PlanImportWorkoutItem) {
  return request.patch<PlanImportDraftRead>(`/plan-imports/${importId}/items/${itemId}`, normalizedItem);
}

export function validatePlanImport(importId: number) {
  return request.post<PlanImportDraftRead>(`/plan-imports/${importId}/validate`);
}

export function applyPlanImport(importId: number) {
  return request.post<{ import_id: number; status: string; diff_summary: unknown }>(`/plan-imports/${importId}/apply`);
}

export function cancelPlanImport(importId: number) {
  return request.post<{ import_id: number; status: string }>(`/plan-imports/${importId}/cancel`);
}

File: server/api/routes/plan_imports.py
from datetime import date

from fastapi import APIRouter, Depends, File, Form, Header, UploadFile
from fastapi.responses import Response
from sqlalchemy.orm import Session

from planner_core.database.models import UserAccount
from server.api.deps import get_current_user, get_db
from server.schemas.plan_import import (
    PlanImportApplyResponse,
    PlanImportCreateResponse,
    PlanImportDraftRead,
    PlanImportItemUpdate,
    PlanImportStructuredRequest,
)
from server.services import plan_import_service, plan_import_template_service

router = APIRouter(prefix="/plan-imports", tags=["plan imports"])


@router.post("/structured", response_model=PlanImportCreateResponse)
def create_structured_plan_import(
    payload: PlanImportStructuredRequest,
    idempotency_key: str | None = Header(default=None, alias="Idempotency-Key"),
    db: Session = Depends(get_db),
    current_user: UserAccount = Depends(get_current_user),
) -> PlanImportCreateResponse:
    return plan_import_service.create_structured_import(db, current_user.id, payload, idempotency_key)


@router.post("/file", response_model=PlanImportCreateResponse)
async def create_file_plan_import(
    file: UploadFile = File(...),
    target_cycle_id: int | None = Form(default=None),
    target_block_id: int | None = Form(default=None),
    client_request_id: str = Form(...),
    anchor_strategy: str = Form(default="after_last_completed"),
    effective_date: date | None = Form(default=None),
    merge_strategy: str = Form(default="replace_uncompleted_in_range"),
    timezone: str = Form(default="Asia/Shanghai"),
    idempotency_key: str | None = Header(default=None, alias="Idempotency-Key"),
    db: Session = Depends(get_db),
    current_user: UserAccount = Depends(get_current_user),
) -> PlanImportCreateResponse:
    content = await file.read()
    return await plan_import_service.create_file_import(
        db,
        current_user.id,
        filename=file.filename or "plan-import",
        content_type=file.content_type,
        content=content,
        target_cycle_id=target_cycle_id,
        target_block_id=target_block_id,
        client_request_id=client_request_id,
        anchor_strategy=anchor_strategy,
        effective_date=effective_date,
        merge_strategy=merge_strategy,
        timezone=timezone,
        idempotency_key=idempotency_key,
    )


@router.get("/template")
def download_plan_import_template(current_user: UserAccount = Depends(get_current_user)) -> Response:
    content = plan_import_template_service.generate_plan_import_template_bytes()
    return Response(
        content=content,
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={"Content-Disposition": 'attachment; filename="plan-import-template.xlsx"'},
    )


@router.get("/{import_id}", response_model=PlanImportDraftRead)
def get_plan_import(
    import_id: int,
    db: Session = Depends(get_db),
    current_user: UserAccount = Depends(get_current_user),
) -> PlanImportDraftRead:
    return plan_import_service.get_plan_import(db, current_user.id, import_id)


@router.patch("/{import_id}/items/{item_id}", response_model=PlanImportDraftRead)
def update_plan_import_item(
    import_id: int,
    item_id: int,
    payload: PlanImportItemUpdate,
    db: Session = Depends(get_db),
    current_user: UserAccount = Depends(get_current_user),
) -> PlanImportDraftRead:
    return plan_import_service.update_plan_import_item(db, current_user.id, import_id, item_id, payload)


@router.post("/{import_id}/validate", response_model=PlanImportDraftRead)
def validate_plan_import(
    import_id: int,
    db: Session = Depends(get_db),
    current_user: UserAccount = Depends(get_current_user),
) -> PlanImportDraftRead:
    return plan_import_service.validate_plan_import(db, current_user.id, import_id)


@router.post("/{import_id}/apply", response_model=PlanImportApplyResponse)
def apply_plan_import(
    import_id: int,
    db: Session = Depends(get_db),
    current_user: UserAccount = Depends(get_current_user),
) -> PlanImportApplyResponse:
    return plan_import_service.apply_plan_import(db, current_user.id, import_id)


@router.post("/{import_id}/cancel", response_model=PlanImportDraftRead)
def cancel_plan_import(
    import_id: int,
    db: Session = Depends(get_db),
    current_user: UserAccount = Depends(get_current_user),
) -> PlanImportDraftRead:
    return plan_import_service.cancel_plan_import(db, current_user.id, import_id)
